Automattic/newspack-plugin
# feat: reorganize segment UI

### All Submissions:






### Changes proposed in this Pull Request:



Reorganizes the UI for editing segments by nesting the options into different "criteria" groups. Criteria can be enabled or disabled. When enabling a criteria group, we check for conflicting criteria and automatically disable any that were enabled. Currently the only conflicting criteria are Reader Engagement and Referrer Sources.

Note that if you have pre-existing segments with conflicting config options (e.g. referrers and min/max articles in session are already set), then this branch won't update those.

<img width="1051" alt="Screen Shot 2021-02-03 at 4 18 40 PM" src="https://user-images.githubusercontent.com/2230142/106822011-84ad8500-663b-11eb-84ec-868559b27ff6.png">

Potentially closes #828. Some questions remain:

* Is this the right approach to address the concerns outlined in #828?
* Is this too blunt an instrument? Referrer source settings only _really_ conflict with session read settings, but in this UI enabling referrer settings will not allow you to set any engagement settings.
* This is a proof-of-concept, so I'd definitely like @thomasguillot to review and make changes as needed (IF we decide this is the right approach to address #828). I tried using only existing components so hopefully it's not a huge stretch.

### How to test the changes in this Pull Request:

1. Check out this branch, run `npm run build`.
2. In the Segments tab, create a new segment and enter a name in the Name field. Observe the new UI with all criteria disabled by default.
3. Enable the Reader Engagement and Reader Activity sections, and change some settings from the default. Confirm that the "Save" and "Cancel" buttons still behave as before (showing a warning if canceling without saving, and enabling the Save button only if the settings have changed from the default or previously saved config).
4. Enable Referrer Sources. Verify that the the Reader Engagement section is automatically disabled and that Reader Activity is unaffected. Enter some values in the Referrer fields.
5. Enable Reader Engagement again. Verify that Referrer Sources is automatically disabled.
6. Confirm that when re-enabling a section that you had previously changed settings for, your previous settings are persisted as long as you don't cancel or save the segment.
7. Ensure that all criteria sections have some fields filled out, even though only one of Reader Engagement or Referrer Sources will be able to be enabled. Note which one is disabled and click "Save" to save your segment. (Any options under disabled sections are discarded on save.)
8. Edit the segment you just saved and confirm that the criteria that had been enabled when you saved are automatically shown as enabled on load.
9. Enable the disabled criteria section and confirm that its config options were reset to the default.
10. Disable all criteria sections and confirm that the Save button is disabled (can't save if no criteria are enabled, which would match the default config).
11. Change some settings in the criteria section that had been disabled on your first save, then disable the section and re-enable the other two sections. Confirm that the Save button is disabled (can't save if the saved config would match the previously saved state, even if there are dirty options in a disabled section).

#### Bonus feature: min/max inputs

As a bonus, I fixed something that bothered me with the existing UI. Now, you can no longer enter a min value that's higher than a max value, and vice versa. To test:

1. Create or edit a segment and enable the Reader Engagement section.
2. Enter a max value for both Articles Read and Articles Read in Session.
3. Attempt to enter min values for both options that are higher than the max values. Confirm that it only lets you enter numbers that are equal to or lower than the max values, whether you use the checkboxes or text inputs.
4. Attempt to lower the max values to something lower than the min values. Confirm that it only lets you enter numbers that are equal to or greater than the min values, whether you use the checkboxes or text inputs.
5. Confirm that if min numbers are disabled or 0, you can enter any number in the max fields, and vice versa.

Ideally I would love to change this UI to something that's more appropriate for a range of numbers, but I'm not really sure what would be best. My first thought was something like the [Material range slider](https://material-ui.com/components/slider/#range-slider), but that would limit the max settings to a concrete number. Definitely open to ideas on this.

### Other information:
